dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/icon/DefaultIconService.java

```diff
@@ -67,33 +71,48 @@ public class DefaultIconService implements IconService {
   private static final String ICON_PATH = "SVGs";
 
   private final IconStore iconStore;
-
   private final FileResourceService fileResourceService;
-
   private final UserService userService;
 
+  private final Set<DefaultIcon> ignoredAfterFailure = ConcurrentHashMap.newKeySet();
+
   @Override
-  @Transactional
-  public void createDefaultIcon(DefaultIcon icon) {
-    if (iconStore.containsKeys(getKeysForVariants(icon))) return;
-
-    for (Icon i : createIconWithVariants(icon)) {
-      if (!iconStore.containsKeys(Set.of(i.getKey()))) {
-        String fileResourceId = CodeGenerator.generateUid();
-        Resource resource = getDefaultIconResource(i.getKey());
-        try {
-          FileResource fileResource =
-              fileResourceService.syncSaveFileResource(
-                  fileResourceId, MediaType.IMAGE_PNG, resource.getInputStream(), CUSTOM_ICON);
-
-          i.setFileResource(fileResource);
-
-          addIconInternal(i);
-        } catch (Exception ex) {
-          throw new RuntimeException(ex);
+  @Transactional(readOnly = true)
+  public List<Icon> findNonExistingDefaultIcons() {
+    Set<String> existing = Set.copyOf(iconStore.getAllKeys());
+    List<Icon> missing = new ArrayList<>();
+    for (DefaultIcon icon : DefaultIcon.values()) {
+      if (!ignoredAfterFailure.contains(icon)
+          && icon.getVariants().stream().anyMatch(key -> !existing.contains(key))) {
+        for (Icon i : icon.createIcons()) {
+          if (!existing.contains(i.getKey())) {
+            missing.add(i);
+          }
         }
       }
     }
+    return missing;
+  }
+
+  @Override
+  @Transactional
+  public String uploadDefaultIcon(Icon icon) throws ConflictException {
+    String fileResourceId = CodeGenerator.generateUid();
+    Resource resource = getDefaultIconResource(icon.getKey());
+    try {
+      FileResource fileResource =
+          FileResource.ofKey(CUSTOM_ICON, icon.getKey(), MediaType.IMAGE_PNG);
```

Comment:
Why is `ofKey` called with `CUSTOM_ICON` in the method `uploadDefaultIcon`? And why `png`? At least right now default icons like `https://play.im.dhis2.org/dev/api/42/icons/clinical_fe_positive/icon` are served as svg 🤔

Reply:
I don't know :) I just kept the arguments as there were earlier just passing them to a new method.

Reply:
Interesting 😅